File: Day-3/PallindromicPartition.js
function isPalindrome(s) {
  let left = 0, right = s.length - 1;
  while (left < right) {
      if (s[left] !== s[right]) return false;
      left++;
      right--;
  }
  return true;
}

function backtrack(s, start, temp, result) {
  if (start === s.length) {
      result.push([...temp]);
      return;
  }

  for (let i = start + 1; i <= s.length; i++) {
      let substr = s.slice(start, i);
      if (isPalindrome(substr)) {
          temp.push(substr);
          backtrack(s, i, temp, result);
          temp.pop();
      }
  }
}

function partition(s) {
  let result = [];
  backtrack(s, 0, [], result);
  return result;
}

// Test cases

File: Day-3/PallindromicPartition.py
def is_palindrome(s: str) -> bool:
    return s == s[::-1]

def backtrack(s: str, start: int, temp: list, result: list):
    if start == len(s):
        result.append(temp[:])
        return

    for i in range(start + 1, len(s) + 1):
        substr = s[start:i]
        if is_palindrome(substr):
            temp.append(substr)
            backtrack(s, i, temp, result)
            temp.pop()

def partition(s: str):
    result = []
    backtrack(s, 0, [], result)
    return result

# Test cases